ext/datadog_cov/datadog_cov.c

```diff
@@ -154,31 +177,43 @@ static void dd_cov_update_coverage(rb_event_flag_t event, VALUE data, VALUE self
 
 static VALUE dd_cov_start(VALUE self)
 {
-  // get current thread
-  VALUE thval = rb_thread_current();
 
   struct dd_cov_data *dd_cov_data;
   TypedData_Get_Struct(self, struct dd_cov_data, &dd_cov_data_type, dd_cov_data);
 
-  if (dd_cov_data->root_len != 0)
+  if (dd_cov_data->root_len == 0)
+  {
+    return self;
+  }
+
+  if (dd_cov_data->threading_mode == SINGLE_THREADED_COVERAGE_MODE)
   {
-    // add event hook
+    VALUE thval = rb_thread_current();
     rb_thread_add_event_hook(thval, dd_cov_update_coverage, RUBY_EVENT_LINE, self);
   }
+  else
+  {
+    rb_add_event_hook(dd_cov_update_coverage, RUBY_EVENT_LINE, self);
+  }
 
   return self;
 }
 
 static VALUE dd_cov_stop(VALUE self)
 {
-  // get current thread
-  VALUE thval = rb_thread_current();
-  // remove event hook for the current thread
-  rb_thread_remove_event_hook(thval, dd_cov_update_coverage);
-
   struct dd_cov_data *dd_cov_data;
   TypedData_Get_Struct(self, struct dd_cov_data, &dd_cov_data_type, dd_cov_data);
 
+  if (dd_cov_data->threading_mode == SINGLE_THREADED_COVERAGE_MODE)
+  {
+    VALUE thval = rb_thread_current();
+    rb_thread_remove_event_hook(thval, dd_cov_update_coverage);
+  }
```

Comment:
So I guess this bug already existed, but it occurred to me that `stop` is incorrect in single-threaded mode if it accidentally gets called on another thread. Aka: maybe store the thread you're currently targeting? (And maybe `raise` if anyone tries to start more than once in single threaded mode?)

Reply:
a-ha, you are right, it is kind of alleviated in the client code as DDCov is created and stored by thread, but I should make it more obvious